fc/parsing/actions.py

```diff
@@ -873,19 +874,80 @@ def lhs_to_sympy(self, variable_generator, number_generator):
+
+    The third and fourth arguments to the call specify the names of the units used for values in the data file columns.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Check syntax is OK - should be ensured by the parser
+        assert len(self.tokens) == 4
+        assert isinstance(self.tokens[0], str)
+        assert isinstance(self.tokens[1], Variable)
+        assert isinstance(self.tokens[2], str)
+        assert isinstance(self.tokens[3], str)
+
+        # Resolve paths relative to the current protocol file
+        self.data_path = os.path.join(os.path.dirname(source_file), self.tokens[0])
+        self.indexing_variable = self.tokens[1]
+        self.index_units = self.tokens[2]
+        self.data_units = self.tokens[3]
+
+    def to_sympy(self, variable_generator, number_generator):
+        """Convert this equation to Sympy.
+
+        Typically this will be called via :meth:`ModelInterface.convert_equations_to_sympy` which sets up appropriate
+        generators.
+
+        :param variable_generator: a function to resolve a name reference within the equation to a model variable.
+        :param number_generator: converts a number with units to a Sympy entity.
+        """
+        index = self.indexing_variable.to_sympy(variable_generator, number_generator)
+        # Load the data from file
+        data_file_name = os.path.basename(self.data_path)
+        if not os.path.exists(self.data_path):
+            raise ProtocolError(f'Unable to load data file {self.data_path} for interpolation.')
+        data = numpy.loadtxt(self.data_path, dtype=float, delimiter=',', ndmin=2, unpack=True)
+        assert data.ndim == 2
+        if data.shape[0] != 2:
+            raise ProtocolError(
+                f'The data file for an interpolate() must have 2 columns; file {data_file_name} has {data.shape[0]}')
+        # Check to see whether we can do a special-case optimisation for equally spaced independent variable points
+        steps = numpy.diff(data[0])
+        if abs(numpy.max(steps) - numpy.min(steps)) < 1e-6 * numpy.max(steps):
+            if steps[0] == 0.0:
+                raise ProtocolError(
+                    f'The data file {data_file_name} used in an interpolate() has all zero table steps.')
+            # Unit-convert the indexing variable if required
+            data_index_units = number_generator(1.0, self.index_units).units
+            index = index.model.convert_variable(
+                index, data_index_units, DataDirectionFlow.OUTPUT, move_annotations=False)
+            # Result is a special Symbol that can generate interpolation code
+            data_units = number_generator(1.0, self.data_units).units
+            return DataInterpolation(data_file_name, data, index, data_units)
+        else:
+            # Iterate over data rows to construct a piecewise representation for the interpolation
```

Comment:
Seems a bit hacky?
Maybe leave for now and add a ticket to add something more sensible in the future, e.g. a `DataInterpolation` variant that finds the correct 2 values to linearly interpolate via bisection, or that replaces the data entirely with a cubic spline?

Reply:
The piecewise is hacky? Much easier than more special-case code though. This is intended for small voltage step protocols. You don't tend to have large numbers of steps I don't think? Certainly not in the samples I've been given when creating this feature!